test/e2e/v4/featuregates_test.go

```diff
@@ -0,0 +1,201 @@
+			Expect(err).NotTo(HaveOccurred())
+			Expect(ctx).NotTo(BeNil())
+
+			// Set up test context
+			ctx.Domain = "example.com"
+			ctx.Group = "crew"
+			ctx.Version = "v1"
+			ctx.Kind = "Captain"
+			ctx.Resources = "captains"
+
+			// Prepare the test environment
+			err = ctx.Prepare()
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		AfterEach(func() {
+			ctx.Destroy()
+		})
+
+		It("should scaffold a project with feature gates", func() {
+			// Initialize the project
+			err := ctx.Init(
+				"--domain", ctx.Domain,
+			)
+			Expect(err).NotTo(HaveOccurred())
+
+			// Create API with feature gates
+			err = ctx.CreateAPI(
+				"--group", ctx.Group,
+				"--version", ctx.Version,
+				"--kind", ctx.Kind,
+				"--resource", "--controller",
+			)
+			Expect(err).NotTo(HaveOccurred())
+
+			// Verify feature gates file was generated
+			featureGatesFile := filepath.Join(ctx.Dir, "internal", "featuregates", "featuregates.go")
+			Expect(featureGatesFile).To(BeAnExistingFile())
+
+			// Verify main.go has feature gates flag
+			mainFile := filepath.Join(ctx.Dir, "cmd", "main.go")
+			Expect(mainFile).To(BeAnExistingFile())
+			mainContent, err := os.ReadFile(mainFile)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(string(mainContent)).To(ContainSubstring("--feature-gates"))
+			Expect(string(mainContent)).To(ContainSubstring("featuregates"))
+
+			// Verify API types have feature gate example
+			typesFile := filepath.Join(ctx.Dir, "api", ctx.Version, strings.ToLower(ctx.Kind)+"_types.go")
+			Expect(typesFile).To(BeAnExistingFile())
+			typesContent, err := os.ReadFile(typesFile)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(string(typesContent)).To(ContainSubstring("+feature-gate"))
+		})
+
+		It("should discover feature gates from API types", func() {
+			// Initialize the project
+			err := ctx.Init(
+				"--domain", ctx.Domain,
+			)
+			Expect(err).NotTo(HaveOccurred())
+
+			// Create API
+			err = ctx.CreateAPI(
+				"--group", ctx.Group,
+				"--version", ctx.Version,
+				"--kind", ctx.Kind,
+				"--resource", "--controller",
+			)
+			Expect(err).NotTo(HaveOccurred())
+
+			// Add custom feature gates to the API types
+			typesFile := filepath.Join(ctx.Dir, "api", ctx.Version, strings.ToLower(ctx.Kind)+"_types.go")
+			typesContent, err := os.ReadFile(typesFile)
+			Expect(err).NotTo(HaveOccurred())
+
+			// Add a new field with feature gate
+			newField := `
+	// Experimental field that requires the "experimental-field" feature gate
+	// +feature-gate experimental-field
+	// +optional
+	ExperimentalField *string ` + "`" + `json:"experimentalField,omitempty"` + "`" + `
+`
+			// Insert the new field in the Spec struct
+			updatedContent := strings.Replace(string(typesContent),
+				"Bar *string `json:\"bar,omitempty\"`",
+				"Bar *string `json:\"bar,omitempty\"`\n"+newField,
+				1)
+
+			err = os.WriteFile(typesFile, []byte(updatedContent), 0644)
+			Expect(err).NotTo(HaveOccurred())
```

Comment:
We have an utility for it: https://github.com/kubernetes-sigs/kubebuilder/blob/master/pkg/plugin/util/util.go#L235